File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        // char[] arr = word.toCharArray();
        // f÷or(int j=0;j<arr.length;j++){
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (helper(board, i, j, word, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    public boolean helper(char[][] board, int i, int j, String word, int index) {

        if (index == word.length()) {
            return true;
        }
        if (i < 0 || i >= board.length || j < 0 || j >= board[i].length || board[i][j] != word.charAt(index)) {
            return false;
        }
        char temp = board[i][j];
        board[i][j]='*';

        boolean found = helper(board, i + 1, j, word, index + 1) ||
                helper(board, i - 1, j, word, index + 1) ||
                helper(board, i, j + 1, word, index + 1) ||
                helper(board, i, j - 1, word, index + 1);
            //         System.out.println(word);
            // return true;
        board[i][j]=temp;
        return found;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        dirs = [(1,0),(0,1),(0,-1),(-1,0)]
        def dfs(i,j,idx):
            if idx == len(word):
                return True
            if i < 0 or j < 0 or i == m or j == n or board[i][j] == "#":
                return False
            if board[i][j] != word[idx]:
                return False
            board[i][j] = "#"
            for dr,dc in dirs:
                nr = dr + i
                nc = dc + j
                if dfs(nr,nc,idx+1):
                    return True
            board[i][j] = word[idx]
        for i in range(m):
            for j in range(n):
                if dfs(i,j,0):
                    return True
        return False

        